src/gc/gceventstatus.h

```diff
@@ -0,0 +1,245 @@
+            break;
+        case GCEventLevel_Warning:
+            fprintf(stderr, "  level: Warning\n");
+            break;
+        case GCEventLevel_Information:
+            fprintf(stderr, "  level: Information\n");
+            break;
+        case GCEventLevel_Verbose:
+            fprintf(stderr, "  level: Verbose\n");
+            break;
+        default:
+            fprintf(stderr, "  level: %d?\n", level);
+            break;
+        }
+
+        fprintf(stderr, "  keywords: ");
+        if (keyword & GCEventKeyword_GC)
+        {
+            fprintf(stderr, "GC ");
+        }
+
+        if (keyword & GCEventKeyword_GCHandle)
+        {
+            fprintf(stderr, "GCHandle ");
+        }
+
+        if (keyword & GCEventKeyword_GCHeapDump)
+        {
+            fprintf(stderr, "GCHeapDump ");
+        }
+
+        if (keyword & GCEventKeyword_GCSampledObjectAllocationHigh)
+        {
+            fprintf(stderr, "GCSampledObjectAllocationHigh ");
+        }
+
+        if (keyword & GCEventKeyword_GCHeapSurvivalAndMovement)
+        {
+            fprintf(stderr, "GCHeapSurvivalAndMovement ");
+        }
+
+        if (keyword & GCEventKeyword_GCHeapCollect)
+        {
+            fprintf(stderr, "GCHeapCollect ");
+        }
+
+        if (keyword & GCEventKeyword_GCHeapAndTypeNames)
+        {
+            fprintf(stderr, "GCHeapAndTypeNames ");
+        }
+
+        if (keyword & GCEventKeyword_GCSampledObjectAllocationLow)
+        {
+            fprintf(stderr, "GCSampledObjectAllocationLow ");
+        }
+
+        fprintf(stderr, "\n");
+    }
+#endif // TRACE_GC_EVENT_STATUS
+
+    // This class is a singleton and can't be instantiated.
+    GCEventStatus() = delete;
+};
+
+class GCEventBase
+{
+private:
+    const char *m_name;
+    GCEventProvider m_provider;
+    GCEventLevel m_level;
+    GCEventKeyword m_keyword;
+
+public:
+    GCEventBase(const char *name, GCEventProvider provider, GCEventLevel level, GCEventKeyword keyword)
+        : m_name(name), m_provider(provider), m_level(level), m_keyword(keyword)
+    {}
+
+    GCEventProvider Provider() const { return m_provider; }
+
+    GCEventLevel Level() const { return m_level; }
+
+    GCEventKeyword Keyword() const { return m_keyword; }
+
+    __forceinline bool IsEnabled() const
+    {
+        return GCEventStatus::IsEnabled(m_provider, m_keyword, m_level);
+    }
+};
+
+class GCKnownEvent : public GCEventBase
+{
+public:
+    GCKnownEvent(const char *name, GCEventProvider provider, GCEventLevel level, GCEventKeyword keyword)
+        : GCEventBase(name, provider, level, keyword)
+    {}
+};
+
+class GCDynamicEvent : public GCEventBase
+{
+    /* TODO(segilles) - Not Yet Implemented */
+};
+
+#if FEATURE_EVENT_TRACE
+#define KNOWN_EVENT(name, _provider, _level, _keyword)   \
+  extern gc_events::GCKnownEvent name##EventDescriptor;
+#include "gcevents.h"
+
+#define EVENT_ENABLED(name) name##EventDescriptor.IsEnabled()
```

Comment:
This has extra level of indirections via the EventDescriptor static. It makes the check slower than it needs to be. The `EVENT_ENABLED` should `IsEnabled(GCEventProvider provider, GCEventKeyword keyword, GCEventLevel level)` directly (with constant arguments).

Instead of EventDescripror statics, you can use the `KNOWN_EVENT` macro to generate constants that the compiler inlines.

Reply:
sure! I figure that some sort of event descriptor strategy will be necessary for dynamic events, but you're right that it's not necessary here.